include/lsst/afw/geom/Angle.h

```diff
@@ -246,73 +299,182 @@ Angle const NullAngle = Angle(-1000000., degrees);
  */
-template<typename T>
-inline
-const Angle operator *(T lhs,              ///< the value to convert
-                       AngleUnit const rhs ///< the conversion coefficient
-                      ) {
-    static_assert(std::numeric_limits<T>::is_specialized,
-                            "Only numeric types may be converted to Angles using degrees/radians!");
-    return Angle(lhs*rhs._val);
+template <typename T>
+inline constexpr Angle operator*(T lhs, AngleUnit rhs) noexcept {
+    static_assert(std::is_arithmetic<T>::value,
+                  "Only numeric types may be converted to Angles using degrees/radians!");
+    return Angle(lhs * rhs._val);
+}
+
+/************************************************************************************************************/
+// Inline method definitions, placed last in order to benefit from Angle's full API
+
+inline double Angle::toUnitSphereDistanceSquared() const noexcept {
+    return 2. * (1. - std::cos(asRadians()));
+    // == 4.0 * pow(std::sin(0.5 * asRadians()), 2.0)
+}
+
+// not constexpr b/c std::acos is not constexpr on OS X
+inline Angle Angle::fromUnitSphereDistanceSquared(double d2) noexcept {
+    return (std::acos(1. - d2 / 2.)) * radians;
+    // == 2.0 * asin(0.5 * sqrt(d2))
+}
+
+inline Angle Angle::wrap() const noexcept {
+    double wrapped = std::fmod(_val, TWOPI);
+    // wrapped is in the range (-TWOPI, TWOPI)
+    if (wrapped < 0.0) wrapped += TWOPI;
+    // if wrapped is small enough, adding 2 pi gives 2 pi
+    if (wrapped >= TWOPI) wrapped = 0.0;
+    return wrapped * radians;
+}
+
+inline Angle Angle::wrapCtr() const noexcept {
+    double wrapped = std::fmod(_val, TWOPI);
+    // wrapped is in the range [-TWOPI, TWOPI]
+    if (wrapped < -PI) {
+        wrapped += TWOPI;
+        if (wrapped >= PI) {
+            // handle roundoff error, however unlikely
+            wrapped = -PI;
+        }
+    } else if (wrapped >= PI) {
+        wrapped -= TWOPI;
+        if (wrapped < -PI) {
+            // handle roundoff error, however unlikely
+            wrapped = -PI;
+        }
+    }
+    return wrapped * radians;
 }
-/**
- * Output operator for an Angle
- */
-std::ostream& operator<<(std::ostream &s, ///< The output stream
-                         Angle const a    ///< The angle
-						 );
 
-}}}
-#endif // if !defined(LSST_AFW_GEOM_ANGLE_H)
+inline Angle Angle::wrapNear(Angle const& refAng) const noexcept {
+    // compute (this - refAng).wrapCtr() + refAng
+    // which is correct except for roundoff error at the edges
+    double const refAngRad = refAng.asRadians();
+    double wrapped = (*this - refAng).wrapCtr().asRadians() + refAngRad;
+
+    // roundoff can cause slightly out-of-range values; fix those
+    if (wrapped - refAngRad >= PI) {
+        wrapped -= TWOPI;
+    }
+    // maximum relative roundoff error for subtraction is 2 epsilon
+    if (wrapped - refAngRad < -PI) {
+        wrapped -= wrapped * 2.0 * std::numeric_limits<double>::epsilon();
```

Comment:
Actually, this code doesn't make sense at all -- it's rescaling `wrapped` by a factor of (1 - 2&epsilon;), but depending on the value of `refAngRad` the absolute value of `wrapped` may be well outside the ~&pi; range. Shouldn't the correction be scaled to `|this - refAng|`?